common/http/src/retry.rs

```diff
@@ -262,22 +284,75 @@ impl RetrySender {
         file.write_all(b"}").await?;
         file.flush().await?;
 
+        let file_size = match metadata(&file_name).await {
+            Ok(md) => md.len(),
+            Err(e) => {
+                debug!("retry file size err, metrics may be skewed; reason={}", e);
+                0
+            }
+        };
+
+        let new_disk_used = if let Some(disk_limit) = self.disk_limit {
+            // Since this block could be executed in multiple async contexts, this code
+            // uses compare and swap constructs to ensure that the check was done on the
+            // most recent value for disk_used. If the update still cannot succeed after
+            // a number of attempts, the code enforces a hard limit and returns an error
+            // to the caller.
+            let mut attempts = 1;
+            loop {
+                let cur_used = self.disk_used.load(SeqCst);
+                let needed = cur_used + file_size;
+                if needed > disk_limit {
+                    warn!(
+                            "retry file not saved; disk limit reached: current={}, required={}, limit={}",
+                            cur_used, needed, disk_limit
+                        );
+                    return Ok(remove_file(file_name).await?);
+                }
+
+                if self
+                    .disk_used
+                    .compare_exchange(cur_used, needed, SeqCst, SeqCst)
+                    .is_ok()
+                {
+                    break needed;
+                } else if attempts > 10 {
+                    remove_file(file_name).await?;
+                    return Err(Error::RetryLimitError(
+                        "failed to update disk_used after 10 attempts".to_string(),
+                    ));
+                }
+
+                attempts += 1;
+            }
+        } else {
+            self.disk_used.fetch_add(file_size, SeqCst);
+            self.disk_used.load(SeqCst)
+        };
+
+        Metrics::retry().report_storage_used(new_disk_used);
+
         let mut new_file_name = self.directory.clone();
         new_file_name.push(format!("{}_{}.retry", fn_ts, fn_uuid));
 
-        Ok(rename(file_name, new_file_name).await?)
+        return Ok(rename(file_name, new_file_name).await?);
```

Comment:
Style nit, unnecessary return?

Reply:
D'OH! Old C habits die hard. Addressed in commit https://github.com/logdna/logdna-agent-v2/pull/279/commits/5795585b171c8a2ad48b1e840ff438a5b0aa72cf.